## Team- 7

Rounak kumar -D22007 ,
Benwin P.A - D22032

In [1]:
#import libraries
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings("ignore")

In [2]:
df_nolabels = pd.read_csv("Loans - Public Test Set (No Labels).csv")
df_labels = pd.read_csv("Loans - Public Test Set (IDs and Labels).csv")

In [3]:
df_nolabels.drop(columns=['Loan Status'],inplace = True)

In [4]:
test_data = pd.merge(left=df_labels,right=df_nolabels,how='inner',on='Loan ID')

In [5]:
test_data.head(5)

Loan ID  Loan Status  \
0  03e0da0c-fac3-42f8-a396-ed5a226466e0   Fully Paid   
1  01825d5d-429a-4de6-b3d4-8618364d2df2  Charged Off   
2  c159a79a-8d2b-447c-92a9-052b5466d6cd   Fully Paid   
3  c1428058-2dc6-4a88-b64c-b6034dfb4f8f  Charged Off   
4  20344186-79bd-487a-b7c4-f3c7b57fe3e8  Charged Off   

                            Customer ID  Current Loan Amount        Term  \
0  16ecb7be-c90f-4bdb-b168-da705d9b1fa7                23023   Long Term   
1  9e0609db-23c1-43b0-9326-0e67ca1e8f1d                 8126  Short Term   
2  c54e404f-1e94-45a2-9473-972e2f3c378b                 5042  Short Term   
3  37923e4e-df73-4093-a4b8-c2dfcbd0c8d8                14139   Long Term   
4  d5dc29d5-06f9-452e-9d55-c1d543dca5dd                15191   Long Term   

   Credit Score Years in current job Home Ownership  Annual Income  \
0           597            10+ years  Home Mortgage          72230   
1           696              2 years           Rent          55869   
2           745            10+ years  Home Mortgage          40332   
3           677             < 1 year           Rent          70693   
4           716            10+ years  Home Mortgage          38442   

              Purpose Monthly Debt  Years of Credit History  \
0  Debt Consolidation     $613.96                      14.0   
1  Debt Consolidation     $665.78                      10.9   
2  Debt Consolidation       873.86                     16.4   
3  Debt Consolidation       482.48                     19.5   
4  Debt Consolidation       611.87                     17.8   

   Months since last delinquent  Number of Open Accounts  \
0                           NaN                       11   
1                           NaN                       14   
2                          11.0                        6   
3                          33.0                        8   
4                           0.0                        4   

   Number of Credit Problems  Current Credit Balance Maximum Open Credit  \
0                          0                   23188                 NaN   
1                          0                    6135                 NaN   
2                          0                    6698             17958.0   
3                          0                    5928              7273.0   
4                          0                   10714             11243.0   

   Bankruptcies  Tax Liens  
0           NaN        NaN  
1           NaN        NaN  
2           0.0        0.0  
3           0.0        0.0  
4           0.0        0.0

In [6]:
test_data.columns

Index(['Loan ID', 'Loan Status', 'Customer ID', 'Current Loan Amount', 'Term',
       'Credit Score', 'Years in current job', 'Home Ownership',
       'Annual Income', 'Purpose', 'Monthly Debt', 'Years of Credit History',
       'Months since last delinquent', 'Number of Open Accounts',
       'Number of Credit Problems', 'Current Credit Balance',
       'Maximum Open Credit', 'Bankruptcies', 'Tax Liens'],
      dtype='object')

In [7]:
test_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 44454 entries, 0 to 44453
Data columns (total 19 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Loan ID                       44454 non-null  object 
 1   Loan Status                   44454 non-null  object 
 2   Customer ID                   44454 non-null  object 
 3   Current Loan Amount           44454 non-null  int64  
 4   Term                          44454 non-null  object 
 5   Credit Score                  44454 non-null  int64  
 6   Years in current job          42604 non-null  object 
 7   Home Ownership                44454 non-null  object 
 8   Annual Income                 44454 non-null  int64  
 9   Purpose                       44454 non-null  object 
 10  Monthly Debt                  44454 non-null  object 
 11  Years of Credit History       44454 non-null  float64
 12  Months since last delinquent  44452 non-null  float64
 13  N

In [8]:
#sum of missing values
sum_total = test_data.isnull().sum().sort_values(ascending=False)

# percentage of missing values
percentage = ((test_data.isnull().sum() / len(test_data))*100).sort_values(ascending=False)


missing_values_df = pd.concat([sum_total,percentage],axis=1,keys = ['sum_total','percent'])
missing_values_df.head(20)

sum_total   percent
Years in current job               1850  4.161605
Tax Liens                             2  0.004499
Bankruptcies                          2  0.004499
Maximum Open Credit                   2  0.004499
Months since last delinquent          2  0.004499
Monthly Debt                          0  0.000000
Current Credit Balance                0  0.000000
Number of Credit Problems             0  0.000000
Number of Open Accounts               0  0.000000
Years of Credit History               0  0.000000
Loan ID                               0  0.000000
Loan Status                           0  0.000000
Annual Income                         0  0.000000
Home Ownership                        0  0.000000
Credit Score                          0  0.000000
Term                                  0  0.000000
Current Loan Amount                   0  0.000000
Customer ID                           0  0.000000
Purpose                               0  0.000000

### As per the result there are missing values columns (Months since last delinquent,Credit Score,Annual Income,Years in current job,Bankruptcies,Tax Liens	).

## Input values for missing value columns and data preprocessing 

In [9]:
# checking the values for the column 'Tax liens'
test_data['Tax Liens'].value_counts()

0.0     43640
1.0       582
2.0       136
3.0        52
4.0        18
5.0        11
6.0         7
7.0         2
9.0         1
10.0        1
8.0         1
13.0        1
Name: Tax Liens, dtype: int64

### Determining the conditional mode for ''Tax Liens'' column

In [10]:
print(test_data[test_data['Loan Status']=='Fully Paid']['Tax Liens'].mode())
print(test_data[test_data['Loan Status']=='Charged Off']['Tax Liens'].mode())

0    0.0
Name: Tax Liens, dtype: float64
0    0.0
Name: Tax Liens, dtype: float64


In [11]:
# Filling the missing values with the mode since it's a categorical column
test_data['Tax Liens'].fillna(value=0.0,inplace=True)

In [12]:
test_data['Tax Liens'].isnull().sum()

0

###  Determining the conditional mode for ''Bankruptcies'' column

In [13]:
test_data['Bankruptcies'].value_counts()

0.0    39809
1.0     4446
2.0      158
3.0       29
4.0        4
5.0        3
7.0        2
6.0        1
Name: Bankruptcies, dtype: int64

In [14]:
# Determining the conditional mode for Bankruptcies column
print(test_data[test_data['Loan Status']=='Fully Paid']['Bankruptcies'].mode())
print(test_data[test_data['Loan Status']=='Charged Off']['Bankruptcies'].mode())

0    0.0
Name: Bankruptcies, dtype: float64
0    0.0
Name: Bankruptcies, dtype: float64


In [15]:
# Filling the missing values with the mode since it's a categorical column
test_data['Bankruptcies'].fillna(value=0.0,inplace=True)

In [16]:
test_data['Bankruptcies'].isnull().sum()

0

###  Determining the conditional mode for ''Years in current job'' column

In [17]:
test_data['Years in current job'].value_counts()

10+ years    13806
2 years       4175
< 1 year      3654
3 years       3485
5 years       3039
1 year        3029
4 years       2717
6 years       2502
7 years       2322
8 years       2147
9 years       1728
Name: Years in current job, dtype: int64

In [18]:
print(test_data[test_data['Loan Status']=='Fully Paid']['Years in current job'].mode())
print(test_data[test_data['Loan Status']=='Charged Off']['Years in current job'].mode())

0    10+ years
Name: Years in current job, dtype: object
0    10+ years
Name: Years in current job, dtype: object


In [19]:
# Filling the missing values with the mode since it's a categorical column
test_data['Years in current job'].fillna(value='10+ years',inplace=True)

In [20]:
test_data['Years in current job'].isnull().sum()

0

### for Home Ownership  

In [21]:
# Replacing every instance of 'HaveMortgage' with 'Home Mortgage'
#test_data['Home Ownership'] = test_data['Home Ownership'].apply(lambda x : 'Home Mortgage' if x == 'HaveMortgage' else x)

In [22]:
test_data['Home Ownership'].value_counts()

Home Mortgage    21742
Rent             18820
Own Home          3892
Name: Home Ownership, dtype: int64

In [23]:
#Can also be done by using loan.replace({'HaveMortgage' : 'Home Mortgage'})

In [24]:
test_data['Home Ownership'].unique()

array(['Home Mortgage', 'Rent', 'Own Home'], dtype=object)

### Credit score

In [25]:
# A value between 0 and 800 indicating the riskiness of the borrowers credit history.
# as per the data dictionary

#df['Credit Score'] = df['Credit Score'].apply(lambda x : x/10 if x > 800 else x)

### checking all other columns

### Current Loan Amount

In [28]:
test_data['Current Loan Amount'].value_counts()

10056    16
8154     16
9825     16
10189    16
9788     15
         ..
12799     1
27230     1
5545      1
15670     1
27618     1
Name: Current Loan Amount, Length: 17946, dtype: int64

In [30]:
test_data['Term'].value_counts()

Short Term    33506
Long Term     10948
Name: Term, dtype: int64

In [31]:
test_data['Credit Score'].value_counts()

740    1349
741    1348
746    1345
747    1344
745    1265
       ... 
590       3
588       3
587       3
592       3
586       1
Name: Credit Score, Length: 167, dtype: int64

In [32]:
test_data['Years in current job'].value_counts()

10+ years    15656
2 years       4175
< 1 year      3654
3 years       3485
5 years       3039
1 year        3029
4 years       2717
6 years       2502
7 years       2322
8 years       2147
9 years       1728
Name: Years in current job, dtype: int64

In [33]:
test_data['Home Ownership'].value_counts()

Home Mortgage    21742
Rent             18820
Own Home          3892
Name: Home Ownership, dtype: int64

In [34]:
test_data['Annual Income'].value_counts()

48970     12
50135     11
50820      9
60570      9
80024      9
          ..
36286      1
50077      1
35473      1
122110     1
34514      1
Name: Annual Income, Length: 27764, dtype: int64

In [35]:
test_data['Purpose'].value_counts()

Debt Consolidation      35315
Home Improvements        2616
other                    2442
Other                    1395
Business Loan             656
Buy a Car                 535
Medical Bills             496
Buy House                 290
Take a Trip               234
major_purchase            150
small_business            141
Educational Expenses       52
moving                     50
vacation                   41
wedding                    38
renewable_energy            3
Name: Purpose, dtype: int64

In [36]:
#in purpose column there are two same levels such as 'other' and 'Other' - replacing them

test_data['Purpose'].replace({'Other':'other'},inplace = True)

In [37]:
test_data['Purpose'].value_counts()

Debt Consolidation      35315
other                    3837
Home Improvements        2616
Business Loan             656
Buy a Car                 535
Medical Bills             496
Buy House                 290
Take a Trip               234
major_purchase            150
small_business            141
Educational Expenses       52
moving                     50
vacation                   41
wedding                    38
renewable_energy            3
Name: Purpose, dtype: int64

### Monthly Income

The monthly debt column had values of object type. But when we tried to convert into 'float' type we encounterd an error due to the presence of '\\$' symbol. Hence we are removing every instance containing
the '$' symbol with a blank character to enable type conversion into integer.

In [38]:
# Note: The datatype of the values are of object type
test_data['Monthly Debt']

0        $613.96 
1        $665.78 
2          873.86
3          482.48
4          611.87
           ...   
44449    1,089.12
44450      509.07
44451    1,224.49
44452      242.24
44453    1,580.01
Name: Monthly Debt, Length: 44454, dtype: object

In [39]:
# Type conversion into integer
#test_data['Monthly Debt'].replace({'\$':'',',':'',' ':''},regex=True,inplace=True)

In [40]:
#test_data["Monthly Debt"] = test_data["Monthly Debt"].replace({"\$":""},regex=True)

In [41]:
test_data["Monthly Debt"] = test_data["Monthly Debt"].replace({"\$":""},regex=True).str.replace(',', '').astype(float)

In [42]:
test_data['Monthly Debt'].value_counts()

0.00       53
426.66      5
852.61      5
502.73      5
782.21      5
           ..
1845.13     1
1745.86     1
130.73      1
412.06      1
1580.01     1
Name: Monthly Debt, Length: 39321, dtype: int64

In [43]:
# Note: The datatype of the values are of float type
test_data['Monthly Debt'].astype('float')

0         613.96
1         665.78
2         873.86
3         482.48
4         611.87
          ...   
44449    1089.12
44450     509.07
44451    1224.49
44452     242.24
44453    1580.01
Name: Monthly Debt, Length: 44454, dtype: float64

In [44]:
test_data['Years of Credit History'].value_counts()

16.0    593
15.0    587
17.0    572
14.0    536
18.0    516
       ... 
43.1      1
54.5      1
49.9      1
49.6      1
54.1      1
Name: Years of Credit History, Length: 496, dtype: int64

In [45]:
test_data['Months since last delinquent'].value_counts()

0.0      24525
12.0       393
10.0       374
7.0        373
9.0        371
         ...  
122.0        1
85.0         1
106.0        1
87.0         1
90.0         1
Name: Months since last delinquent, Length: 104, dtype: int64

In [46]:
test_data['Number of Open Accounts'].value_counts().sort_index(ascending = True)

0        1
1        5
2      187
3      570
4     1278
5     2128
6     2934
7     3532
8     3969
9     4255
10    4077
11    3710
12    3277
13    2751
14    2319
15    1915
16    1565
17    1296
18    1015
19     809
20     619
21     471
22     412
23     319
24     245
25     167
26     146
27     112
28      69
29      65
30      61
31      32
32      33
33      22
34      16
35      15
36      10
37       8
38      14
39       7
40       3
41       3
42       4
43       1
45       3
46       3
50       1
Name: Number of Open Accounts, dtype: int64

In [47]:
test_data['Number of Credit Problems'].value_counts(0)

0     38572
1      5173
2       488
3       129
4        41
5        27
6        14
8         4
7         3
9         1
10        1
13        1
Name: Number of Credit Problems, dtype: int64

In [48]:
test_data['Current Credit Balance'].value_counts()

0        293
7507       9
4831       9
6612       9
4943       8
        ... 
31120      1
20468      1
4750       1
13747      1
43782      1
Name: Current Credit Balance, Length: 24578, dtype: int64

In [49]:
test_data['Maximum Open Credit'].value_counts()

0.0        240
0          104
9512.0       7
12168.0      7
9817.0       6
          ... 
21593.0      1
50311.0      1
28968.0      1
13128.0      1
67462        1
Name: Maximum Open Credit, Length: 34984, dtype: int64

In [50]:
(test_data['Maximum Open Credit']=='0').sum()

104

In [51]:
(test_data['Maximum Open Credit']==0.0).sum()

240

We observe that there are zeros in the column with string as well as integer datatype. So we're converting all the values of this variable into numeric.

In [52]:
test_data['Maximum Open Credit'] = pd.to_numeric(test_data['Maximum Open Credit'],errors='coerce')

#### Can also be done using the following code
loan['Maximum Open Credit'].apply(lambda x: float(x) if type(x)==str() else x,inplace=True)

In [53]:
test_data['Maximum Open Credit'].value_counts().sort_index()

0.0            344
261.0            1
293.0            2
297.0            2
298.0            2
              ... 
10556549.0       1
12081044.0       1
12379714.0       1
15605108.0       1
176332155.0      1
Name: Maximum Open Credit, Length: 30641, dtype: int64

In [54]:
# conveting the monthly debt column from object data type to numeric data type
test_data['Monthly Debt'] = pd.to_numeric(test_data['Monthly Debt'])

### Determining the conditional mode for ''Annual Income'' column

In [55]:
#fully_paid = df[df['Loan Status']=='Fully Paid']['Annual Income']
#charged_off = df[df['Loan Status']=='Charged Off']['Annual Income']
#fully_paid.fillna(value=fully_paid.median(),inplace=True)
#charged_off.fillna(value=charged_off.median(),inplace=True)

In [56]:
#df['Annual Income'][df['Loan Status']=='Fully Paid'] = fully_paid

#df['Annual Income'][df['Loan Status']=='Charged Off'] = charged_off

#df['Annual Income'].isnull().sum()

### Determining the conditional mode for ''Credit Score'' column

In [57]:
#fully_paid = df[df['Loan Status']=='Fully Paid']['Credit Score']
#charged_off = df[df['Loan Status']=='Charged Off']['Credit Score']
#fully_paid.fillna(value=fully_paid.median(),inplace=True)
#charged_off.fillna(value=charged_off.median(),inplace=True)

In [58]:
#df['Credit Score'][df['Loan Status']=='Fully Paid'] = fully_paid

#df['Credit Score'][df['Loan Status']=='Charged Off'] = charged_off

#df['Credit Score'].isnull().sum()

### Determining the conditional mode for ''Months since last delinquent' column

In [59]:
fully_paid = test_data[test_data['Loan Status']=='Fully Paid']['Months since last delinquent']
charged_off = test_data[test_data['Loan Status']=='Charged Off']['Months since last delinquent']
fully_paid.fillna(value=fully_paid.median(),inplace=True)
charged_off.fillna(value=charged_off.median(),inplace=True)

In [60]:
test_data['Months since last delinquent'][test_data['Loan Status']=='Fully Paid'] = fully_paid

test_data['Months since last delinquent'][test_data['Loan Status']=='Charged Off'] = charged_off

test_data['Months since last delinquent'].isnull().sum()


0

In [61]:
test_data.isnull().sum() / len(test_data)*100

Loan ID                         0.000000
Loan Status                     0.000000
Customer ID                     0.000000
Current Loan Amount             0.000000
Term                            0.000000
Credit Score                    0.000000
Years in current job            0.000000
Home Ownership                  0.000000
Annual Income                   0.000000
Purpose                         0.000000
Monthly Debt                    0.000000
Years of Credit History         0.000000
Months since last delinquent    0.000000
Number of Open Accounts         0.000000
Number of Credit Problems       0.000000
Current Credit Balance          0.000000
Maximum Open Credit             0.006749
Bankruptcies                    0.000000
Tax Liens                       0.000000
dtype: float64

In [62]:
test_data['Maximum Open Credit'].isnull().sum()

3

In [63]:
#observed 2 missing values 
test_data.drop(index= list(test_data[test_data['Maximum Open Credit'].isnull()].index),inplace=True)

In [64]:
test_data['Maximum Open Credit'].isnull().sum()

0

In [71]:
test_data.isnull().sum() / len(test_data)*100

Loan ID                         0.0
Loan Status                     0.0
Customer ID                     0.0
Current Loan Amount             0.0
Term                            0.0
Credit Score                    0.0
Years in current job            0.0
Home Ownership                  0.0
Annual Income                   0.0
Purpose                         0.0
Monthly Debt                    0.0
Years of Credit History         0.0
Months since last delinquent    0.0
Number of Open Accounts         0.0
Number of Credit Problems       0.0
Current Credit Balance          0.0
Maximum Open Credit             0.0
Bankruptcies                    0.0
Tax Liens                       0.0
dtype: float64

## Checking for duplicate variables 

In [72]:
# Determining the number of duplicate observations corresponding to 
test_data.shape[0]-test_data['Loan ID'].nunique()

0

In [73]:
# dropping the duplicate observations with respect to 'LOAN ID' column
test_data.drop_duplicates('Loan ID',inplace=True)

In [74]:
# Determining the number of duplicate observations after drop duplicates  
test_data.shape[0]-test_data['Loan ID'].nunique()

0

## Using Loan Status as my target variable 

In [75]:
test_data['Loan Status'].value_counts()

1    31867
0    12584
Name: Loan Status, dtype: int64

### it is found that it has imbalanced data,so converting it into balanced data.

In [76]:
test_data['Loan Status'].replace({'Charged Off':0,'Fully Paid':1},inplace = True)

In [77]:
test_data['Term'].replace({'Short Term':0,'Long Term':1},inplace = True)

In [78]:
test_data.head(3)

Loan ID  Loan Status  \
2  c159a79a-8d2b-447c-92a9-052b5466d6cd            1   
3  c1428058-2dc6-4a88-b64c-b6034dfb4f8f            0   
4  20344186-79bd-487a-b7c4-f3c7b57fe3e8            0   

                            Customer ID  Current Loan Amount  Term  \
2  c54e404f-1e94-45a2-9473-972e2f3c378b                 5042     0   
3  37923e4e-df73-4093-a4b8-c2dfcbd0c8d8                14139     1   
4  d5dc29d5-06f9-452e-9d55-c1d543dca5dd                15191     1   

   Credit Score Years in current job Home Ownership  Annual Income  \
2           745            10+ years  Home Mortgage          40332   
3           677             < 1 year           Rent          70693   
4           716            10+ years  Home Mortgage          38442   

              Purpose  Monthly Debt  Years of Credit History  \
2  Debt Consolidation        873.86                     16.4   
3  Debt Consolidation        482.48                     19.5   
4  Debt Consolidation        611.87                     17.8   

   Months since last delinquent  Number of Open Accounts  \
2                          11.0                        6   
3                          33.0                        8   
4                           0.0                        4   

   Number of Credit Problems  Current Credit Balance  Maximum Open Credit  \
2                          0                    6698              17958.0   
3                          0                    5928               7273.0   
4                          0                   10714              11243.0   

   Bankruptcies  Tax Liens  
2           0.0        0.0  
3           0.0        0.0  
4           0.0        0.0

In [79]:
test_data['Years in current job'].replace({'< 1 year':0,'1 year':1,'2 years':2,'3 years':3,'4 years':4,
                                        '5 years':5,'6 years':6,'7 years':7,'8 years':8,'9 years':9,
                                        '10+ years':10},inplace = True)

In [80]:
test_data.head(5)

Loan ID  Loan Status  \
2  c159a79a-8d2b-447c-92a9-052b5466d6cd            1   
3  c1428058-2dc6-4a88-b64c-b6034dfb4f8f            0   
4  20344186-79bd-487a-b7c4-f3c7b57fe3e8            0   
5  5b85d870-178d-40dd-bc91-bc48392f6572            1   
6  b352edea-9c4d-4e68-9253-97e07142fc32            1   

                            Customer ID  Current Loan Amount  Term  \
2  c54e404f-1e94-45a2-9473-972e2f3c378b                 5042     0   
3  37923e4e-df73-4093-a4b8-c2dfcbd0c8d8                14139     1   
4  d5dc29d5-06f9-452e-9d55-c1d543dca5dd                15191     1   
5  5547bfc2-bc90-4802-927b-eda98859a247                10023     1   
6  ca8a978e-f50e-4390-ad0f-dafd10da7c46                 2524     0   

   Credit Score  Years in current job Home Ownership  Annual Income  \
2           745                    10  Home Mortgage          40332   
3           677                     0           Rent          70693   
4           716                    10  Home Mortgage          38442   
5           724                     0           Rent          40092   
6           733                    10  Home Mortgage          26199   

              Purpose  Monthly Debt  Years of Credit History  \
2  Debt Consolidation        873.86                     16.4   
3  Debt Consolidation        482.48                     19.5   
4  Debt Consolidation        611.87                     17.8   
5               other        544.58                     18.4   
6   Home Improvements        240.16                     24.2   

   Months since last delinquent  Number of Open Accounts  \
2                          11.0                        6   
3                          33.0                        8   
4                           0.0                        4   
5                           0.0                        7   
6                           0.0                        4   

   Number of Credit Problems  Current Credit Balance  Maximum Open Credit  \
2                          0                    6698              17958.0   
3                          0                    5928               7273.0   
4                          0                   10714              11243.0   
5                          1                    3126              13711.0   
6                          1                     533               2961.0   

   Bankruptcies  Tax Liens  
2           0.0        0.0  
3           0.0        0.0  
4           0.0        0.0  
5           1.0        0.0  
6           1.0        0.0

In [81]:
test_data.drop(columns=['Loan ID', 'Customer ID'],inplace = True)

In [82]:
test_data.head()

Loan Status  Current Loan Amount  Term  Credit Score  Years in current job  \
2            1                 5042     0           745                    10   
3            0                14139     1           677                     0   
4            0                15191     1           716                    10   
5            1                10023     1           724                     0   
6            1                 2524     0           733                    10   

  Home Ownership  Annual Income             Purpose  Monthly Debt  \
2  Home Mortgage          40332  Debt Consolidation        873.86   
3           Rent          70693  Debt Consolidation        482.48   
4  Home Mortgage          38442  Debt Consolidation        611.87   
5           Rent          40092               other        544.58   
6  Home Mortgage          26199   Home Improvements        240.16   

   Years of Credit History  Months since last delinquent  \
2                     16.4                          11.0   
3                     19.5                          33.0   
4                     17.8                           0.0   
5                     18.4                           0.0   
6                     24.2                           0.0   

   Number of Open Accounts  Number of Credit Problems  Current Credit Balance  \
2                        6                          0                    6698   
3                        8                          0                    5928   
4                        4                          0                   10714   
5                        7                          1                    3126   
6                        4                          1                     533   

   Maximum Open Credit  Bankruptcies  Tax Liens  
2              17958.0           0.0        0.0  
3               7273.0           0.0        0.0  
4              11243.0           0.0        0.0  
5              13711.0           1.0        0.0  
6               2961.0           1.0        0.0

In [83]:
test_data = pd.get_dummies(test_data)
test_data

Loan Status  Current Loan Amount  Term  Credit Score  \
2                1                 5042     0           745   
3                0                14139     1           677   
4                0                15191     1           716   
5                1                10023     1           724   
6                1                 2524     0           733   
...            ...                  ...   ...           ...   
44449            0                12292     0           735   
44450            1                 4931     0           743   
44451            0                23272     1           632   
44452            0                18611     1           668   
44453            1                27618     1           721   

       Years in current job  Annual Income  Monthly Debt  \
2                        10          40332        873.86   
3                         0          70693        482.48   
4                        10          38442        611.87   
5                         0          40092        544.58   
6                        10          26199        240.16   
...                     ...            ...           ...   
44449                    10          60506       1089.12   
44450                     9          34514        509.07   
44451                    10          68663       1224.49   
44452                    10          44312        242.24   
44453                     0          93400       1580.01   

       Years of Credit History  Months since last delinquent  \
2                         16.4                          11.0   
3                         19.5                          33.0   
4                         17.8                           0.0   
5                         18.4                           0.0   
6                         24.2                           0.0   
...                        ...                           ...   
44449                     18.3                          51.0   
44450                     13.0                           0.0   
44451                     15.5                           0.0   
44452                     31.8                          24.0   
44453                     23.9                           0.0   

       Number of Open Accounts  ...  Purpose_Home Improvements  \
2                            6  ...                          0   
3                            8  ...                          0   
4                            4  ...                          0   
5                            7  ...                          0   
6                            4  ...                          1   
...                        ...  ...                        ...   
44449                       13  ...                          0   
44450                        8  ...                          0   
44451                       16  ...                          0   
44452                       11  ...                          0   
44453                       14  ...                          0   

       Purpose_Medical Bills  Purpose_Take a Trip  Purpose_major_purchase  \
2                          0                    0                       0   
3                          0                    0                       0   
4                          0                    0                       0   
5                          0                    0                       0   
6                          0                    0                       0   
...                      ...                  ...                     ...   
44449                      0                    0                       0   
44450                      0                    0                       0   
44451                      0                    0                       0   
44452                      0                    0                       0   
44453                      0                    0                       0   

       Purpose_moving  Purpose_other  Purp

In [84]:
# defining predicted and target variables
X=test_data.drop(['Loan Status'], axis=1)
y=test_data['Loan Status']

## Train - Test split

In [85]:
#For test-train split

from sklearn.model_selection import train_test_split

# To handle the target class imbalance we make use of stratified random sampling
X_train,X_test,y_train,y_test = train_test_split(X,y,stratify=y)

In [86]:
X_train

Current Loan Amount  Term  Credit Score  Years in current job  \
34894                19682     0           747                    10   
44440                18337     0           748                     1   
33130                 9772     0           716                    10   
31973                12374     0           728                    10   
26696                 6648     0           743                     9   
...                    ...   ...           ...                   ...   
18170                23959     1           746                    10   
2515                 19225     1           717                    10   
7807                 16662     0           744                     9   
8810                 25428     0           747                     5   
31058                15267     0           743                     9   

       Annual Income  Monthly Debt  Years of Credit History  \
34894          72991        875.89                     22.7   
44440         132431       1082.62                     25.0   
33130          40065       1031.68                     23.0   
31973          45995       1238.05                     18.5   
26696          42665        351.27                     17.0   
...              ...           ...                      ...   
18170         109813        323.03                     20.5   
2515           71848       1413.00                     15.4   
7807           89189       1152.03                     24.4   
8810          152565        939.55                     28.8   
31058          71246        213.74                     18.0   

       Months since last delinquent  Number of Open Accounts  \
34894                           0.0                        6   
44440                          35.0                       12   
33130                          63.0                       14   
31973                          10.0                       14   
26696                           0.0                       13   
...                             ...                      ...   
18170                           0.0                        9   
2515                           43.0                        6   
7807                           26.0                        4   
8810                            0.0                       20   
31058                          43.0                        4   

       Number of Credit Problems  ...  Purpose_Home Improvements  \
34894                          0  ...                          0   
44440                          0  ...                          0   
33130                          0  ...                          0   
31973                          0  ...                          0   
26696                          1  ...                          0   
...                          ...  ...                        ...   
18170                          0  ...                          1   
2515                           0  ...                          0   
7807                           0  ...                          0   
8810                           0  ...                          0   
31058                          0  ...                          0   

       Purpose_Medical Bills  Purpose_Take a Trip  Purpose_major_purchase  \
34894                      0                    0                       0   
44440                      0                    0                       0   
33130                      0                    0                       0   
31973                      0                    0                       0   
26696                      0                    0                       0   
...                      ...                  ...                     ...   
18170                      0                    0                       0   
2515                       0                    0                       0   
7807                       0                    0                       0   
8810                       0 

## using SMOTE method to balance the data

In [88]:
from imblearn.over_sampling import SMOTE
sm = SMOTE()
X_res, y_res = sm.fit_resample(X_train, y_train)

In [89]:
y_res.value_counts()

1    23900
0    23900
Name: Loan Status, dtype: int64

### target variables has been balanced using SMOTE

## Inputing MODELS

In [90]:
from sklearn.tree import DecisionTreeClassifier 
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.ensemble import BaggingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV,train_test_split 
from sklearn.metrics import roc_auc_score,f1_score,accuracy_score,classification_report

## Decision Tree

In [91]:
# importing decision tree classifier model (baseline model - fully grown tree)
dtc = DecisionTreeClassifier()
dtc

DecisionTreeClassifier()

In [92]:
# fit the model
dtc.fit(X_res, y_res)
# train the model 
y_pred = dtc.predict(X_test)

In [93]:
print('Accuracy score: ',accuracy_score(y_test,y_pred))
print('Area under ROC :',roc_auc_score(y_test,y_pred))
print('F1-score: ',f1_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

Accuracy score:  0.6227841266984613
Area under ROC : 0.5660151805472846
F1-score:  0.7259414225941423
              precision    recall  f1-score   support

           0       0.36      0.44      0.40      3146
           1       0.76      0.70      0.73      7967

    accuracy                           0.62     11113
   macro avg       0.56      0.57      0.56     11113
weighted avg       0.65      0.62      0.63     11113



## Random Forest

In [95]:
rfc = RandomForestClassifier()

In [96]:
rfc.fit(X_res, y_res)
y_pred = rfc.predict(X_test)

In [97]:
print('Accuracy score: ',accuracy_score(y_test,y_pred))
print('Area under ROC :',roc_auc_score(y_test,y_pred))
print('F1-score: ',f1_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

Accuracy score:  0.7152884009718348
Area under ROC : 0.612450767393885
F1-score:  0.8105389221556887
              precision    recall  f1-score   support

           0       0.50      0.38      0.43      3146
           1       0.77      0.85      0.81      7967

    accuracy                           0.72     11113
   macro avg       0.64      0.61      0.62     11113
weighted avg       0.70      0.72      0.70     11113



## Adaptive boosting

In [98]:
abc = AdaBoostClassifier()

In [99]:
abc.fit(X_res, y_res)
y_pred = abc.predict(X_test)

In [100]:
print('Accuracy score: ',accuracy_score(y_test,y_pred))
print('Area under ROC :',roc_auc_score(y_test,y_pred))
print('F1-score: ',f1_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

Accuracy score:  0.6902726536488797
Area under ROC : 0.620008803798185
F1-score:  0.7835492390894228
              precision    recall  f1-score   support

           0       0.45      0.46      0.46      3146
           1       0.79      0.78      0.78      7967

    accuracy                           0.69     11113
   macro avg       0.62      0.62      0.62     11113
weighted avg       0.69      0.69      0.69     11113



## Bagging

In [104]:
bag = BaggingClassifier()

In [105]:
bag.fit(X_res, y_res)
y_pred = bag.predict(X_test)

In [106]:
print('Accuracy score: ',accuracy_score(y_test,y_pred))
print('Area under ROC :',roc_auc_score(y_test,y_pred))
print('F1-score: ',f1_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

Accuracy score:  0.6621074417349051
Area under ROC : 0.6057509676557568
F1-score:  0.7573819215610261
              precision    recall  f1-score   support

           0       0.42      0.48      0.44      3146
           1       0.78      0.74      0.76      7967

    accuracy                           0.66     11113
   macro avg       0.60      0.61      0.60     11113
weighted avg       0.68      0.66      0.67     11113



## KNN

In [107]:
knn = KNeighborsClassifier()

In [108]:
knn.fit(X_res, y_res)
y_pred = knn.predict(X_test)

In [109]:
print('Accuracy score: ',accuracy_score(y_test,y_pred))
print('Area under ROC :',roc_auc_score(y_test,y_pred))
print('F1-score: ',f1_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

Accuracy score:  0.5701430756771348
Area under ROC : 0.551517220869207
F1-score:  0.6647484033967297
              precision    recall  f1-score   support

           0       0.33      0.51      0.40      3146
           1       0.75      0.59      0.66      7967

    accuracy                           0.57     11113
   macro avg       0.54      0.55      0.53     11113
weighted avg       0.63      0.57      0.59     11113



## logistic

In [110]:
lg = LogisticRegression(solver='newton-cg', class_weight='balanced')

In [111]:
lg.fit(X_res, y_res)
y_pred = lg.predict(X_test)

In [112]:
print('Accuracy score: ',accuracy_score(y_test,y_pred))
print('Area under ROC :',roc_auc_score(y_test,y_pred))
print('F1-score: ',f1_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

Accuracy score:  0.7145685233510303
Area under ROC : 0.5812694784932537
F1-score:  0.8169648009232545
              precision    recall  f1-score   support

           0       0.49      0.27      0.35      3146
           1       0.76      0.89      0.82      7967

    accuracy                           0.71     11113
   macro avg       0.62      0.58      0.58     11113
weighted avg       0.68      0.71      0.69     11113



## COMMENT

### used model such as Knn classifier,decision tree classifier,random forest classifier ,Adaptive boosting,Logistic regression 

### among all the model 
### ADAPTIVE BOOSTING is getting higher Area under ROC score : 0.62